lib/sport_ngin_aws_auditor/instance_helper.rb

```diff
@@ -109,18 +79,82 @@ def sort_through_RIs(client)
       return ris_region, ris_hash
     end
 
-    def measure_differences(instance_hash, ris_hash)
+    def measure_differences(instance_hash, ris_hash, ris_region, klass)
       differences = Hash.new()
+
+      # Setup and zone-based RI calculation
       instance_hash.keys.concat(ris_hash.keys).uniq.each do |key|
         instance_count = instance_hash.has_key?(key) ? instance_hash[key][:count] : 0
         ris_count = ris_hash.has_key?(key) ? ris_hash[key][:count] : 0
-        differences[key] = {:count => ris_count - instance_count, :region_based => false}
+        # positive count => more ris, negative count => more instances
+        differences[key] = { count: ris_count - instance_count, region_based: false }
       end
+
+      # Region-based RI calculation
+      ris_region_processor(differences, ris_region) unless ris_region.empty?
+
       differences
     end
 
+    # Assuming zone based information is already added and differences hash is initialized
+    private def ris_region_processor(differences, ris_region)
+      # Group all the same size RIs
+      # Ex: ri_group_arr = [{ instance_type: "t2.medium", platform: "Linux VPC", count: 7 }, ...]
+      ri_group_arr = group_ris_region(ris_region)
+
+      # Process each ri_group determined by instance_type and platform
+      ri_group_arr.each do |ri_group|
+        target_instance_type, target_platform, ri_count = ri_group[:instance_type], ri_group[:platform], ri_group[:count]
+        instances_left = differences.select do |k,v|
+          # Ex: k,v = "Linux VPC us-east-1d t2.small", {:count=>-15, :region_based=>false}
+          Regexp.new("^#{target_platform}.*#{target_instance_type}$") =~ k
+        end
+
+        instances_left_count = - instances_left.reduce(0) do | previous_count, actual_instance |
```

Comment:
What does this line do? Specifically the `- instances.left.reduce(0)`?

Reply:
`instances_left_count` is the number of instances left after we do the math with AZ specific instances. Let's say there are
```
Two AZ specific t2.small RIs for us-east-1a
Two AZ specific t2.small RIs for us-east-1b
Three t2.small instances in us-east-1a
Four t2.small instances in us-east-1b
```

Then, `instances_left_count` will be 3 because one `t2.small instance in us-east-1a` and two `t2.small instances in us-east-1b` are left.

Reply:
What does the `-` after the `=` mean?

Reply:
When there are more instances than AZ specific RIs , `instances_left` looks like 
```
{
"Linux VPC us-east-1d t2.small"=>{:count=>-3, :region_based=>false}, 
"Linux VPC us-east-1b t2.small"=>{:count=>-5, :region_based=>false}, 
"Windows VPC us-east-1b t2.small"=>{:count=>-1, :region_based=>false}, 
"Linux VPC us-east-1b t2.medium"=>{:count=>-5, :region_based=>false}
}
```
because the way we have historically calculated the differences https://github.com/sportngin/sport_ngin_aws_auditor/blob/beccd7b310c9109cb1bf4375ef01407eed998148/lib/sport_ngin_aws_auditor/instance_helper.rb#L90